### PR TITLE
validation of SWIM image importing, golden tagging status of image, i…

### PR DESCRIPTION
**Pull request by Abhishek-121**
…mage distribution on single or list of devices, image activation on single or list of devices, code for updating management IP address with some dump ip.

In this commit - 

— Write the code for validation of SWIM image importing, golden tagging status of image, image distribution on single or list of devices, image activation on single or list of devices.

— Added the code for updating management IP address with some dump ip retaining the previous value.

-- Add seperate API for each swim verify operation(Import/Tagging/Distribution/Activation) validation.

### CODE REVIEW

File: plugins/modules/inventory_intent.py [hunk 1]
```diff
@@ -40,7 +40,7 @@
     required: True
     suboptions:
       cli_transport:
-        description: Device's cli transport protocol. Required for Adding Network Devices.
+        description: Device's cli transport protocol(ssh/telnet). Required for Adding Network Devices.
```
Comment (madhansansel) [inline plugins/modules/inventory_intent.py:43]: Can we say like below?
To add Network Devices, it's important to mention the way the device communicates, which can be either through SSH or Telnet.
(or)
The essential prerequisite for adding Network Devices is the specification of the transport protocol (either SSH or Telnet) used by the device

File: plugins/modules/inventory_intent.py [hunk 2]
```diff
@@ -2567,6 +2567,37 @@ def get_diff_merged(self, config):
...
+                            if response and isinstance(response, dict):
+                                task_id = response.get('response').get('taskId')
+
+                                while True:
+                                    execution_details = self.get_task_details(task_id)
+                                    if execution_details.get("isError"):
+                                        self.status = "failed"
+                                        failure_reason = execution_details.get("failureReason")
+                                        if failure_reason:
+                                            self.msg = "Device Updation get failed because of {0}".format(failure_reason)
```
Comment (madhansansel) [inline plugins/modules/inventory_intent.py:2588]: Failed to update device due to {0}".format(failure_reason) Also specify which device .. may be device ip

File: plugins/modules/inventory_intent.py [hunk 3]
```diff
@@ -2567,6 +2567,37 @@ def get_diff_merged(self, config):
...
+                                        else:
+                                            self.msg = "Device Updation get failed"
```
Comment (madhansansel) [inline plugins/modules/inventory_intent.py:2590]: Failed to update device " Also specify which device .. may be device ip

File: plugins/modules/inventory_intent.py [hunk 4]
```diff
@@ -2567,6 +2567,37 @@ def get_diff_merged(self, config):
...
+                                        self.log(self.msg, "ERROR")
+                                        break
+                                    elif execution_details.get("endTime"):
+                                        self.status = "success"
+                                        self.result['changed'] = True
+                                        self.result['response'] = execution_details
+                                        self.msg = "Devices {0} present in Cisco Catalyst Center and updated successfully".format(str(device_to_update))
```
Comment (madhansansel) [inline plugins/modules/inventory_intent.py:2597]: "Device(s) {0} are present in the Cisco Catalyst Center and have been updated successfully".format(str(device_to_update))

File: plugins/modules/inventory_intent.py [hunk 5]
```diff
@@ -2598,6 +2622,14 @@ def get_diff_merged(self, config):
                                         self.msg = "Device Updation get failed"
                                     self.log(self.msg, "ERROR")
                                     break
+                                elif execution_details.get("endTime"):
+                                    self.status = "success"
+                                    self.result['changed'] = True
+                                    self.result['response'] = execution_details
+                                    self.msg = "Devices {0} present in Cisco Catalyst Center and updated successfully".format(str(device_to_update))
```
Comment (madhansansel) [inline plugins/modules/inventory_intent.py:2629]: "Devices {0} are present in the Cisco Catalyst Center and have been updated successfully".format(str(device_to_update))

File: plugins/modules/swim_intent.py [hunk 6]
```diff
@@ -1362,6 +1379,125 @@ def get_diff_merged(self, config):
...
+            image_exist = self.is_image_exist(name)
+            if image_exist:
+                self.status = "success"
+                self.msg = "The requested Image '{0}' imported in the Cisco Catalyst Center and Image presence has been verified.".format(name)
+                self.log(self.msg, "INFO")
+            else:
+                self.log("""The playbook input for SWIM Image '{0}' does not align with the Cisco Catalyst Center, indicating that image
+                         may not have imported successfully.""".format(name), "INFO")
+
+        if self.want.get("tagging_details"):
```
Comment (madhansansel) [inline plugins/modules/swim_intent.py:1419]: Can we write separate APIs?
```
if import_type:
    self.verify_diff_imported(import_type)

tagged = self.want.get("tagging_details");
if tagged:
    self.verify_diff_tagged(tagged)

distribution_details = self.want.get("distribution_details")
if distribution_details:
    self.verify_diff_distributed(distribution_details)

activation_details = self.want.get("activation_details")
if activation_details:
    self.verify_diff_activated()
```

File: plugins/modules/swim_intent.py [hunk 7]
```diff
@@ -1362,6 +1379,125 @@ def get_diff_merged(self, config):
...
+                params=image_params
+            )
+            self.log("Received API response from 'get_golden_tag_status_of_an_image': {0}".format(str(response)), "DEBUG")
+
+            response = response.get('response')
+            if response:
+                image_status = response['taggedGolden']
+                if image_status == tag_image_golden:
+                    if tag_image_golden:
+                        self.msg = """The requested image '{0}' has been tagged as golden in the Cisco Catalyst Center and image status
```
Comment (madhansansel) [inline plugins/modules/swim_intent.py:1444]: can we say successfully verified?
For example:
The requested image '{0}' has been tagged as golden in the Cisco Catalyst Center and its status has been successfully verified.".format(name)

File: plugins/modules/swim_intent.py [hunk 8]
```diff
@@ -1362,6 +1379,125 @@ def get_diff_merged(self, config):
...
+            else:
+                self.log("""Mismatch between the playbook input for tagging/un-tagging image as golden and the Cisco Catalyst Center indicates that
+                         the tagging/un-tagging task was not executed successfully.""", "INFO")
+
+        if self.want.get("distribution_details"):
+            image_id = self.have.get("distribution_image_id")
+
+            if self.have.get("distribution_device_id"):
+                if self.single_device_distribution:
+                    self.msg = """The requested image with device id '{0}' has been distributed successfully in the Cisco Catalyst Center and
```
Comment (madhansansel) [inline plugins/modules/swim_intent.py:1460]: The requested image, associated with the device ID '{0}', has been successfully distributed in the Cisco Catalyst Center and its status has been verified.".format(self.have.get("distribution_device_id"))

File: plugins/modules/swim_intent.py [hunk 9]
```diff
@@ -1362,6 +1379,125 @@ def get_diff_merged(self, config):
...
+                             image status has been verified.""".format(self.have.get("distribution_device_id"))
+                    self.log(self.msg, "INFO")
+                else:
+                    self.log("""Mismatch between the playbook input for distributing image to device with id '{0}' and the Cisco Catalyst Center indicates that
```
Comment (madhansansel) [inline plugins/modules/swim_intent.py:1464]: Mismatch between the playbook input for distributing the image to the device with ID '{0}' and the actual state in the Cisco Catalyst Center suggests that the distribution task might not have been executed successfully.

File: plugins/modules/swim_intent.py [hunk 10]
```diff
@@ -1362,6 +1379,125 @@ def get_diff_merged(self, config):
...
+                         the distribution task was not executed successfully.""".format(self.have.get("distribution_device_id")), "INFO")
+            elif self.complete_successful_distribution:
+                self.msg = """The requested image with id '{0}' distributed successfully on all devices in that particular site
```
Comment (madhansansel) [inline plugins/modules/swim_intent.py:1467]: The requested image, with ID '{0}', has been successfully distributed to all devices within the specified site in the Cisco Catalyst Center.".format(image_id)

Can we specify the site?
Can we print the devices (may be IP addresses) ?

File: plugins/modules/swim_intent.py [hunk 11]
```diff
@@ -1362,6 +1379,125 @@ def get_diff_merged(self, config):
...
+                         in the Cisco Catalyst Center.""".format(image_id)
+                self.log(self.msg, "INFO")
+            elif self.partial_successful_distribution:
+                self.msg = "The requested image with id '{0}' distributed partially on some devices in the Cisco Catalyst Center.".format(image_id)
```
Comment (madhansansel) [inline plugins/modules/swim_intent.py:1471]: "The requested image, with ID '{0}', has been partially distributed across some devices in the Cisco Catalyst Center.".format(image_id)

Also can we specify which devices?

File: plugins/modules/swim_intent.py [hunk 12]
```diff
@@ -1362,6 +1379,125 @@ def get_diff_merged(self, config):
...
+                self.log(self.msg, "INFO")
+            else:
+                self.msg = "The requested image with id '{0}' failed to distributed on devices in the Cisco Catalyst Center.".format(image_id)
```
Comment (madhansansel) [inline plugins/modules/swim_intent.py:1474]: The requested image, with ID '{0}', failed to be distributed across devices in the Cisco Catalyst Center.".format(image_id)

Also which devices???

File: plugins/modules/swim_intent.py [hunk 13]
```diff
@@ -1362,6 +1379,125 @@ def get_diff_merged(self, config):
...
+                self.log(self.msg, "INFO")
+
+        if self.want.get("activation_details"):
+            image_id = self.have.get("activation_image_id")
+
+            if self.have.get("activation_device_id"):
+                if self.single_device_activation:
+                    self.msg = """The requested image with device id '{0}' has been activated successfully in the Cisco Catalyst Center and
```
Comment (madhansansel) [inline plugins/modules/swim_intent.py:1482]: The requested image, associated with the device ID '{0}', has been successfully activated in the Cisco Catalyst Center and its status has been verified.".format(self.have.get("activation_device_id"))

Print the image name???

File: plugins/modules/swim_intent.py [hunk 14]
```diff
@@ -1362,6 +1379,125 @@ def get_diff_merged(self, config):
...
+                             image status has been verified.""".format(self.have.get("activation_device_id"))
+                    self.log(self.msg, "INFO")
+                else:
+                    self.log("""Mismatch between the playbook input for activating image to device with id '{0}' and the Cisco Catalyst Center indicates that
```
Comment (madhansansel) [inline plugins/modules/swim_intent.py:1486]: print image name??

Mismatch between the playbook's input for activating the image on the device with ID '{0}' and the actual state in the Cisco Catalyst Center suggests that the activation task might not have been executed successfully.".format(self.have.get("activation_device_id"))

File: plugins/modules/swim_intent.py [hunk 15]
```diff
@@ -1362,6 +1379,125 @@ def get_diff_merged(self, config):
...
+                         the activation task was not executed successfully.""".format(self.have.get("activation_device_id")), "INFO")
+            elif self.complete_successful_activation:
+                self.msg = """The requested image with id '{0}' activated successfully on all devices in that particular site
```
Comment (madhansansel) [inline plugins/modules/swim_intent.py:1489]: The requested image, with ID '{0}', has been successfully activated on all devices within the specified site in the Cisco Catalyst Center.".format(image_id)

image name is more important than image id?

File: plugins/modules/swim_intent.py [hunk 16]
```diff
@@ -1362,6 +1379,125 @@ def get_diff_merged(self, config):
...
+                         in the Cisco Catalyst Center.""".format(image_id)
+                self.log(self.msg, "INFO")
+            elif self.partial_successful_activation:
+                self.msg = "The requested image with id '{0}' activated partially on some devices in the Cisco Catalyst Center.".format(image_id)
+                self.log(self.msg, "INFO")
+            else:
+                self.msg = "The requested image with id '{0}' failed to activate on devices in the Cisco Catalyst Center.".format(image_id)
```
Comment (madhansansel) [inline plugins/modules/swim_intent.py:1496]: The activation of the requested image, with ID '{0}', failed on devices in the Cisco Catalyst Center.".format(image_id)



image name is more important.. Is it possible to list the devices?

File: plugins/modules/swim_intent.py [hunk 17]
```diff
@@ -1362,6 +1379,125 @@ def get_diff_merged(self, config):
...
+                    self.log(self.msg, "INFO")
+                else:
+                    self.log("""Mismatch between the playbook input for activating image to device with id '{0}' and the Cisco Catalyst Center indicates that
+                         the activation task was not executed successfully.""".format(self.have.get("activation_device_id")), "INFO")
+            elif self.complete_successful_activation:
+                self.msg = """The requested image with id '{0}' activated successfully on all devices in that particular site
+                         in the Cisco Catalyst Center.""".format(image_id)
+                self.log(self.msg, "INFO")
+            elif self.partial_successful_activation:
+                self.msg = "The requested image with id '{0}' activated partially on some devices in the Cisco Catalyst Center.".format(image_id)
```
Comment (madhansansel) [inline plugins/modules/swim_intent.py:1493]: "The requested image '{1}', with ID '{0}', has been partially activated on some devices in the Cisco Catalyst Center.".format(image_id, image_name)

image name is more important.. Is it possible to list the devices?

File: plugins/modules/swim_intent.py (hunk 6 above)
Comment (Abhishek-121) [reply plugins/modules/swim_intent.py:1419]: added seperate API for each swim operation.

File: plugins/modules/swim_intent.py (hunk 7 above)
Comment (Abhishek-121) [reply plugins/modules/swim_intent.py:1444]: added.

File: plugins/modules/swim_intent.py (hunk 8 above)
Comment (Abhishek-121) [reply plugins/modules/swim_intent.py:1460]: added.

File: plugins/modules/swim_intent.py (hunk 9 above)
Comment (Abhishek-121) [reply plugins/modules/swim_intent.py:1464]: added.

File: plugins/modules/swim_intent.py (hunk 10 above)
Comment (Abhishek-121) [reply plugins/modules/swim_intent.py:1467]: For now I added the message, and we are printing device ips while executing the process. But here it'll be little complicated so will take once all pending items get done.

File: plugins/modules/swim_intent.py (hunk 11 above)
Comment (Abhishek-121) [reply plugins/modules/swim_intent.py:1471]: For now I added the message, and we are printing device ips while executing the process. But here it'll be little complicated so will take once all pending items get done.

File: plugins/modules/swim_intent.py (hunk 12 above)
Comment (Abhishek-121) [reply plugins/modules/swim_intent.py:1474]: For now I added the message, and we are printing device ips while executing the process. But here it'll be little complicated so will take once all pending items get done.

File: plugins/modules/swim_intent.py (hunk 14 above)
Comment (Abhishek-121) [reply plugins/modules/swim_intent.py:1486]: added the image name.

File: plugins/modules/swim_intent.py (hunk 15 above)
Comment (Abhishek-121) [reply plugins/modules/swim_intent.py:1489]: added.

File: plugins/modules/swim_intent.py (hunk 16 above)
Comment (Abhishek-121) [reply plugins/modules/swim_intent.py:1496]: added the image name, but listing the device ips can;t be done. Although I'm logging the device ip's while performing the activation operation.

File: plugins/modules/swim_intent.py (hunk 17 above)
Comment (Abhishek-121) [reply plugins/modules/swim_intent.py:1493]: added the image name, but listing the device ips can;t be done. Although I'm logging the device ip's while performing the activation operation.

**Review by madhansansel — APPROVED** [review]

File: plugins/modules/inventory_intent.py (hunk 1 above)
Comment (Abhishek-121) [reply plugins/modules/inventory_intent.py:43]: added.

File: plugins/modules/inventory_intent.py (hunk 3 above)
Comment (Abhishek-121) [reply plugins/modules/inventory_intent.py:2590]: added the log message with device ip.

File: plugins/modules/inventory_intent.py (hunk 2 above)
Comment (Abhishek-121) [reply plugins/modules/inventory_intent.py:2588]: added the log message with device ip.

File: plugins/modules/inventory_intent.py (hunk 5 above)
Comment (Abhishek-121) [reply plugins/modules/inventory_intent.py:2629]: added the log message with device ip.

File: plugins/modules/inventory_intent.py (hunk 4 above)
Comment (Abhishek-121) [reply plugins/modules/inventory_intent.py:2597]: added.

**Pull request merged**